Simulator/Utils/Factory.h

```diff
@@ -0,0 +1,225 @@
+#include "Vertices/NG911/All911Vertices.h"
+#include "Vertices/Neuro/AllIZHNeurons.h"
+#include "Vertices/Neuro/AllLIFNeurons.h"
+
+// Recorder
+#include "Recorders/IRecorder.h"
+#include "Recorders/NG911/Xml911Recorder.h"
+#include "Recorders/Neuro/XmlGrowthRecorder.h"
+#include "Recorders/Neuro/XmlSTDPRecorder.h"
+#include "Recorders/XmlRecorder.h"
+
+#if defined(HDF5)
+   #include "Recorders/Hdf5Recorder.h"
+   #include "Recorders/Neuro/Hdf5GrowthRecorder.h"
+#endif
+
+// MTRand
+#include "RNG/MTRand.h"
+#include "RNG/Norm.h"
+
+template <typename T> class Factory {
+public:
+   // Default destructor
+   ~Factory() = default;
+
+   /// Acts as constructor, returns the instance of singleton object for the instantiated type.
+   static Factory &getInstance()
+   {
+      /** 
+       * Why the T* argument?
+       * C++ functions resolve polymorphic behavior based on their input arguments,
+       * not their return types. This is because the compiler generates a separate
+       * instance of the function for each function signature. 
+       * 
+       * For the CreateFunctionMap, the T* argument is only used to resolve polymorphic behavior.
+       * Also, you can create a pointer, T*, without having to create an instance of the class T. 
+       * In this case,it is essential, because the function expects a pointer to the (abstract) 
+       * base class.
+       */
+
+      T *placeholder = nullptr;
+      static Factory instance(CreateFunctionMap(placeholder));
+      return instance;
+   }
+
+   // Invokes constructor for desired concrete class
+   ///
+   /// @param className class name.
+   /// @return Unique pointer to an instance if className is found in
+   ///         createFunctions map, nullptr otherwise.
+   std::unique_ptr<T> createType(const std::string &className)
+   {
+      auto createIter = createFunctions.find(className);
+      if (createIter != createFunctions.end()) {
+         return std::unique_ptr<T>(createIter->second());
+      }
+      return nullptr;
+   }
+
+   /// Delete copy and move methods to avoid copy instances of the singleton
+   Factory(const Factory &factory) = delete;
+   Factory &operator=(const Factory &factory) = delete;
+
+   Factory(Factory &&factory) = delete;
+   Factory &operator=(Factory &&factory) = delete;
+
+private:
+   /// Defines function type for usage in internal map
+   using CreateFunction = T *(*)(void);
+
+   /// Defines map between class name and corresponding ::Create() function.
+   using FunctionMap = std::map<std::string, CreateFunction>;
+
+   /// Makes class-to-function map an internal factory member.
+   FunctionMap createFunctions;
+
+   /// Constructor is private to keep a singleton instance of this class.
+   Factory(std::map<std::string, CreateFunction> map)
+   {
+      createFunctions = std::move(map);
+   }
+
+   static std::map<std::string, T *(*)(void)> CreateFunctionMap(T *placeholder)
+   {
+      std::map<std::string, T *(*)(void)> createFunctionMap;
```

Comment:
`T *(*)(void)` can be replaced with `CreateFunction`. Also, as indicated in a reply to another comment, I think that this method no longer needs a parameter.

BTW, was `Factory.cpp` done away with? I actually think that's OK in this case.

Reply:
> T *(*)(void) can be replaced with CreateFunction.
 
Agreed.

> was Factory.cpp done away with? 

I was getting linker errors while trying to make the method static, and it would only resolve by adding Factory.cpp in Factory.h file. Creating a circular dependency. So decided to remove it altogether

Reply:
Any tips on resolving the code style violations? `clang-format -i Simulator/Utils/Factory.h`  does not work but `clang-format -style=llvm -i Simulator/Utils/Factory.h` works, but the changes made are still not accepted by GitHub checks

Reply:
Did you try running it manually from the command line? If you do that, you can have it output a modified version of the file that is "fixed". You can then `diff` it with the original to see what it changed (and to check it didn't damage anything, just to be paranoid).

Reply:
Yeah, I did try it, but unfortunately, it didn't help.

Reply:
I remember we upgraded the clang-format on our machines to the latest one. But when I check for the version it is still the older version 
```
dkamath@otachi:~/Graphitti$ clang-format -version
clang-format version 7.0.1-8+deb10u2 (tags/RELEASE_701/final)

```

Reply:
The latest version is available on `csslab13.uwb.edu`.

Reply:
Thank you this suggestion helped!